neqo-http3/src/recv_message.rs

```diff
@@ -337,115 +345,93 @@ impl RecvMessage {
-            let is_colon_header = header.0.starts_with(':');
-            if is_colon_header {
-                if seen_non_colon_header {
-                    qdebug!([self], "headers_valid - pseudo header fields after fields");
-                    return Err(Error::HttpGeneralProtocolStream);
-                }
+            let is_pseudo = self.track_pseudo(&header.0, &mut pseudo_state)?;
 
-                if let hash_map::Entry::Occupied(mut e) = pseudo_header_map.entry(&header.0) {
-                    if *e.get_mut() {
-                        return Err(Error::HttpGeneralProtocolStream);
-                    }
-                    *e.get_mut() = true;
-                } else {
+            let mut bytes = header.0.bytes();
+            if is_pseudo {
+                if (pseudo_state & (PSEUDO_HEADER_UNDEFINED | REGULAR_HEADER)) != 0 {
                     return Err(Error::HttpGeneralProtocolStream);
                 }
 
                 if header.0 == ":method" {
-                    method_value = Some(&header.1)
+                    method_value = Some(&header.1);
                 }
-
-                // skip the first ":"
-                header_byte_iterator.next();
-            } else {
-                seen_non_colon_header = true;
+                let _ = bytes.next();
             }
 
-            for byte in header_byte_iterator {
-                if byte.is_ascii_uppercase() {
-                    qdebug!(
-                        [self],
-                        "headers_valid - found uppercase in header:'{}'",
-                        header.0
-                    );
-                    return Err(Error::HttpGeneralProtocolStream);
-                }
-
-                if excluded_header_byte
-                    .iter()
-                    .any(|&excluded| excluded == byte)
-                {
-                    qdebug!([self], "headers_valid - found an excluded byte");
-                    return Err(Error::HttpGeneralProtocolStream);
-                }
+            if bytes.any(|b| matches!(b, 0 | 0x10 | 0x13 | 0x3a | 0x41..=0x5a)) {
+                return Err(Error::HttpGeneralProtocolStream); // illegal characters.
             }
 
-            if excluded_headers
-                .iter()
-                .any(|&excluded| excluded == header.0)
-            {
-                qdebug!(
-                    [self],
-                    "headers_valid - found an excluded header:'{}'",
-                    header.0
-                );
+            if matches!(
+                header.0.as_str(),
+                "connection"
+                    | "host"
+                    | "keep-alive"
+                    | "proxy-connection"
+                    | "te"
+                    | "transfer-encoding"
+                    | "upgrade"
+                    | "accept-encoding"
+            ) {
                 return Err(Error::HttpGeneralProtocolStream);
             }
         }
 
+        pseudo_state ^= REGULAR_HEADER;
```

Comment:
```suggestion
        pseudo_state &= !REGULAR_HEADER;
```

If you want to clear the bit, then this is safer.  Your code relies on there being a regular header (which seems likely, but not guaranteed).